src/gflownet/utils/focus_model.py

```diff
@@ -0,0 +1,103 @@
+        self.device = device
+        self.n_objectives = n_objectives
+        self.state_space_res = state_space_res
+        self.focus_cosim = focus_cosim
+
+    def update_belief(self, focus_dirs: torch.Tensor, flat_rewards: torch.Tensor):
+        raise NotImplementedError
+
+    def sample_focus_directions(self, n):
+        raise NotImplementedError
+
+
+class TabularFocusModel(FocusModel):
+    """
+    Tabular model of the feasibility of focus directions for goal-condtioning.
+    We keep a count of the number of times each focus direction has been sampled and whether
+    this direction succesfully lead to a sample in this region of the objective space. The (unormalized) likelihood
+    of a focus direction being feasible is then given by the ratio of these numbers.
+    If a focus direction has not been sampled yet it obtains the maximum likelihood of one.
+    """
+
+    def __init__(self, device, n_objectives: int, state_space_res: int, focus_cosim: float) -> None:
+        super().__init__(device, n_objectives, state_space_res, focus_cosim)
+        self.n_objectives = n_objectives
+        self.state_space_res = state_space_res
+        self.focus_dir_dataset = (
+            nn.functional.normalize(torch.tensor(get_limits_of_hypercube(n_objectives, state_space_res)), dim=1)
+            .float()
+            .to(self.device)
+        )
+        self.focus_dir_count = torch.zeros(self.focus_dir_dataset.shape[0]).to(self.device)
+        self.focus_dir_population_count = torch.zeros(self.focus_dir_dataset.shape[0]).to(self.device)
+
+    def update_belief(self, focus_dirs: torch.Tensor, flat_rewards: torch.Tensor):
+        """
+        Updates the focus model with the focus directions and rewards
+        of the last batch.
+        """
+        focus_dirs = nn.functional.normalize(focus_dirs, dim=1)
+        flat_rewards = nn.functional.normalize(flat_rewards, dim=1)
+
+        focus_dirs_indices = torch.argmin(torch.cdist(focus_dirs, self.focus_dir_dataset), dim=1)
+        flat_rewards_indices = torch.argmin(torch.cdist(flat_rewards, self.focus_dir_dataset), dim=1)
+
+        for idxs, count in zip(
+            [focus_dirs_indices, flat_rewards_indices],
+            [self.focus_dir_count, self.focus_dir_population_count],
+        ):
+            idx_increments = torch.bincount(idxs, minlength=len(count))
+            count += idx_increments
+
+    def sample_focus_directions(self, n):
+        """
+        Samples n focus directions from the focus model.
+        """
+        sampling_likelihoods = torch.zeros_like(self.focus_dir_count).float().to(self.device)
+        sampling_likelihoods[self.focus_dir_count == 0] = 1.0
+        sampling_likelihoods[torch.logical_and(self.focus_dir_count > 0, self.focus_dir_population_count > 0)] = 1.0
+        sampling_likelihoods[torch.logical_and(self.focus_dir_count > 0, self.focus_dir_population_count == 0)] = 0.1
```

Comment:
This `0.1` looks like a hyperparameter, would be good to at least be tied to an attribute of the class.

Reply:
Makes sense, I created a new attribute for it to make it more explicit. I haven't wired it in the hps as I don't think it makes a big difference for now, but it could be advisable to add it eventually in the future in a case where the number of infeasible directions would largely outnumber the number of feasible directions (e.g. in massively multi-objective cases).